src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -330,129 +371,131 @@ int MPL_gavl_tree_free(MPL_gavl_tree_t gavl_tree)
+                STACK_PUSH(node_stack, cur_node);
+                cur_node = cur_node->right;
+                continue;
+            } else {
+                *cur_node_ret = NULL;
+                break;
+            }
+        } else {
+            /* find matched node to be deleted */
+            *cur_node_ret = cur_node;
+            break;
+        }
+    } while (1);
+
+    return;
+}
+
+static void gavl_tree_delete_node(gavl_tree_s * gavl_tree_iptr, gavl_tree_node_s * cur_node,
+                                  gavl_tree_node_s ** cur_node_ret, void **val,
+                                  gavl_tree_stack_s * node_stack)
+{
+    gavl_tree_node_s *inorder_node;
+
+    if (cur_node->right == NULL) {
+        /* no right child, next inorder node is parent */
+        if (cur_node->parent == NULL) {
+            /* delete root node */
+            if (cur_node->left) {
+                gavl_tree_iptr->root = cur_node->left;
+                gavl_tree_iptr->root->parent = NULL;
+            } else {
+                gavl_tree_iptr->root = NULL;
+            }
+            *val = (void *) cur_node->val;
+            MPL_free(cur_node);
+        } else {
+            /* assign deleted node's left child to its parent */
+            inorder_node = cur_node->parent;
+            if (inorder_node->left == cur_node)
+                inorder_node->left = cur_node->left;
+            else
+                inorder_node->right = cur_node->left;
+            if (cur_node->left)
+                cur_node->left->parent = inorder_node;
+
+            *val = (void *) cur_node->val;
+            MPL_free(cur_node);
+            STACK_PUSH(node_stack, inorder_node);
+        }
+    } else {
+        /* find next inorder node and replace deleted node with it */
+        inorder_node = cur_node->right;
+        STACK_PUSH(node_stack, cur_node);
+        while (inorder_node->left) {
+            STACK_PUSH(node_stack, inorder_node);
+            inorder_node = inorder_node->left;
+        }
+
+        if (inorder_node->parent != cur_node) {
+            if (inorder_node->right)
+                inorder_node->right->parent = inorder_node->parent;
+            inorder_node->parent->left = inorder_node->right;
+        } else {
+            /* only right child of deleted node */
+            cur_node->right = NULL;
+        }
+        *val = (void *) cur_node->val;
+        cur_node->addr = inorder_node->addr;
+        cur_node->len = inorder_node->len;
+        cur_node->val = inorder_node->val;
+        MPL_free(inorder_node);
+    }
+
+    STACK_POP(node_stack, gavl_tree_node_s *, *cur_node_ret);
+    return;
+}
+
 int MPL_gavl_tree_delete(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t len)
 {
     int mpl_err = MPL_SUCCESS;
     gavl_tree_node_s *cur_node;
-    gavl_tree_node_s *inorder_node;
     gavl_tree_s *gavl_tree_iptr = (gavl_tree_s *) gavl_tree;
 
-    DECLARE_STACK(gavl_tree_node_s *, node_stack);
+    DECLARE_STACK(node_stack);
```

Comment:
I see. If you hide the node_stack inside the tree, then you don't need declare them.